elastic/kibana
# [Discover] Add support for unmapped fields using the fields API

## Summary
Fixes https://github.com/elastic/kibana/issues/88192.
This PR adds support for unmapped fields using the new fields API. The logic we discussed to implement this is as follows:

1. There is a switch in the UI with the deprecation notice; the switch is on by default for existing saved searches; off by default for new saved searches
2. To determine if it's an existing (pre 7.12) saved search, we need a data migration


Current flow:
(notice that for same index pattern unmapped fields don't show if it's a new saved search, but they show up when you load a pre-existing one)
![unmapped_fields](https://user-images.githubusercontent.com/1937956/105643316-0c9ace80-5e87-11eb-8f82-9cf04712fdc2.gif)

#### How to test this?
1. Before pulling the branch, create an index in ES with unmapped fields. Eg:
```
PUT /my-test-index
{
  "settings": {
    "number_of_shards": 1
  },
  "mappings": {
    "properties": {
      "field1": { "type": "text" }
    }
  }
}

PUT /my-test-index/_doc/1
{
  "field1": "Some blob of text",
  "field2": "Some other text"
}

PUT /my-test-index/_doc/2
{
  "field1": "Some blob of text",
  "field2": "Some other text"
}

PUT /my-test-index/_doc/3
{
  "field1": "Some blob of text",
  "field3": 12
}

PUT /my-test-index-2
{
  "settings": {
    "number_of_shards": 1
  },
  "mappings": {
     "dynamic": false
  }
}

PUT /my-test-index-2/_doc/1
{
  "field1": "Some blob of text",
  "field2": "Some other text"
}

PUT /my-test-index-2/_doc/2
{
  "field1": "Some blob of text",
  "field2": "Some other text"
}

PUT /my-test-index-2/_doc/3
{
  "field4": "Some blob of text",
  "field5": 12
}


PUT /my-test-index-2/_doc/4
{
  "field4": "Some blob of text",
  "field5": 12
}
```
2. Create an index pattern from the newly created index
3. Go to Discover, select the index pattern created under 2
4. Save the existing saved search
5. Pull the branch, restart Kibana 
6. Observe how unmapped fields are behaving

### Checklist

Delete any items that are not applicable to this PR.

- [ ] If a plugin configuration key changed, check if it needs to be whitelisted in the [cloud](https://github.com/elastic/cloud) and added to the [docker list](https://github.com/elastic/kibana/blob/c29adfef29e921cc447d2a5ed06ac2047ceab552/src/dev/build/tasks/os_packages/docker_generator/resources/bin/kibana-docker)

### For maintainers
